### PR TITLE
[CSL-2398] Integration tests in CI

### PR DESCRIPTION
**Pull request by akegalj**
Still very much WIP. Description will be expanded when it is close to be finished.

### CODE REVIEW

File: wallet-new/integration/Types.hs [hunk 1]
```diff
@@ -81,7 +81,7 @@ data Action
 -- TODO(ks): We could create a custom constructor with valid
 -- values only.
 -- Add invariant?
-type ActionProbabilities = [(Action, Probability)]
+type ActionProbabilities = NonEmpty (Action, Probability)
```
Comment (parsonsmatt) [inline wallet-new/integration/Types.hs]: :+1: nice invariant :)

**Comment by parsonsmatt** [comment]
I'm unable to get this working okay. I've pushed my changes up to the PR so far, but I am getting 500 errors inexplicably. I don't know how to view or understand the log output either :\

**Comment by akegalj** [comment]
> I'm unable to get this working okay. I've pushed my changes up to the PR so far, but I am getting 500 errors inexplicably. I don't know how to view or understand the log output either :\

@parsonsmatt I have managed to track where to problem is https://iohk.myjetbrains.com/youtrack/issue/CSL-2430 .

File: scripts/test/wallet/integration.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+# wait until cluster is fully up and running
+echo "Waiting 40 seconds until local cluster is ready..."
+sleep 40s
+
+# import keys
+echo "Importing poor HD key/wallet..."
+ curl -X POST http://localhost:8090/api/wallets/keys -H 'cache-control: no-cache' -H 'content-type: application/json' -d "\"$tmpSecrets/generated-keys/poor/key0.sk\""
+
+FAILED=0
```
Comment (parsonsmatt) [inline scripts/test/wallet/integration.sh]: tfw no `finally` in bash

File: wallet-new/integration/CLI.hs [hunk 3]
```diff
@@ -9,22 +9,22 @@ module CLI
...
+    tlsPubCertPath <- pure "./scripts/tls-files/server.crt"
+--        <- strOption $
+--           long        "tls-pub-cert"
+--        <> metavar     "PUB_CERT"
+--        <> help        "Path to TLS public certificate"
+    tlsPrivKeyPath <- pure "./scripts/tls-files/server.key"
+--        <- strOption $
+--           long        "tls-priv-key"
+--        <> metavar     "PRIV_KEY"
+--        <> help        "Path to TLS private key"
```
Comment (parsonsmatt) [inline wallet-new/integration/CLI.hs]: We need to actually take these as parameters next. I suppose we can use these as default values?

File: wallet-new/integration/Error.hs [hunk 4]
```diff
@@ -21,51 +22,64 @@ import           Cardano.Wallet.Client (ClientError)
...
 
     | AddressBalanceNotZero WalletAddress
     | LocalAddressesDiffer WalletAddress [WalletAddress]
-    | LocalAddressDiffer Address
+    | LocalAddressDiffer Address Address
 
     | InvalidTransactionState Transaction
     | InvalidTransactionFee EstimatedFees
-    | LocalTransactionsDiffer [Transaction]
+    | LocalTransactionsDiffer [Transaction] [Transaction]
```
Comment (parsonsmatt) [inline wallet-new/integration/Error.hs]: The error constructors now contain an `expected` and `actual` value, where a diff can be present.

File: wallet-new/integration/Functions.hs [hunk 5]
```diff
@@ -21,467 +25,476 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
...
     -> ActionProbabilities
     -> m WalletState
 runActionCheck walletClient walletState actionProb = do
-    action <- chooseAction actionProb
-    runAction walletClient walletState action
-
+    actions <- toList <$> chooseActions 20 actionProb
+    let client' = hoistClient lift walletClient
+    execStateT (tryAll (map (runAction client') actions)) walletState
+        `catch` \x -> fmap (const walletState) . liftIO . hspec .
```
Comment (parsonsmatt) [inline wallet-new/integration/Functions.hs]: the `hspec` bit is *really* nice for getting an informative error message on why the test failed. kinda want to use hspec to do all the testing, so we get the "prettier" output + can hook it up into the deterministic tests

File: wallet-new/server/Cardano/Wallet/Server/Plugins.hs [hunk 6]
```diff
@@ -110,15 +115,45 @@ legacyWalletBackend WalletBackendParams {..} ntpStatus =
...
+    conv (err, msg, txt) =
+        err { Servant.errBody = encode $ object
+                [ "status" .= ("error" :: Text)
+                , "diagnostic" .= object
+                    [ "msg" .= txt
+                    ]
+                , "message" .= msg
+                ]
+            , Servant.errHeaders = applicationJson : Servant.errHeaders err
+            }
```
Comment (parsonsmatt) [inline wallet-new/server/Cardano/Wallet/Server/Plugins.hs]: This one *might* want to be factored out into a separate PR and story.

**Comment by parsonsmatt** [comment]
I am fairly consistently getting an error like this:

```
[TEST-LOG] Running: UpdateWallet
[TEST-LOG] Updating wallet: WalletId "J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg", WalletUpdate {uwalAssuranceLevel = NormalAssurance, uwalName = "Wallet name WalletId \"J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg\""}
[TEST-LOG] Old   : Wallet {walId = WalletId "J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg", walName = "My wallet", walBalance = Coin {getCoin = 20147218287599424}, walHasSpendingPassword = False, walSpendingPasswordLastUpdate = 1298390588298360, walCreatedAt = 1283517260179742, walAssuranceLevel = NormalAssurance}
[TEST-LOG] Result: Wallet {walId = WalletId "J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg", walName = "My wallet", walBalance = Coin {getCoin = 37138290651997550}, walHasSpendingPassword = False, walSpendingPasswordLastUpdate = 308128273141538, walCreatedAt = 249785469591062, walAssuranceLevel = NormalAssurance}
[TEST-LOG] Success!
[TEST-LOG] Running: PostWallet
[TEST-LOG] Request: NewWallet {newwalBackupPhrase = <backup phrase>, newwalSpendingPassword = Just <passphrase>, newwalAssuranceLevel = NormalAssurance, newwalName = "Wallet", newwalOperation = CreateWallet}
[TEST-LOG] Mnemonic was already taken!
[TEST-LOG] Success!
[TEST-LOG] Running: GetTransaction
[TEST-LOG] Running: GetWallet
[TEST-LOG] Requesting wallet: WalletId "J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg"
cardano-integration-test: FailureResponse (Response {responseStatusCode = Status {statusCode = 400, statusMessage = "Bad Request"}, responseBody = "{\"status\":\"error\",\"diagnostic\":{\"msg\":\"No wallet with address J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg found\"},\"message\":\"BadRequest\"}", responseHeaders = fromList [("Transfer-Encoding","chunked"),("Date","Fri, 06 Apr 2018 22:54:31 GMT"),("Server","Warp/3.2.13"),("Content-Type","application/json")], responseHttpVersion = HTTP/1.1})
```

We update the wallet, and that succeeds, but later when we go to get it,
the wallet does not exist. Not sure what that is about.

**Comment by adinapoli-iohk** [comment]
@parsonsmatt Miscellanea braindump for you:

1. A genuine bug lying around, likely to be similar to the CSL-2430, where we were updating the `wallet-db` but not re-reading the updated `WalletSnapshot`;
2. We didn't merge into this branch the hotfix for [CSL-2430](https://github.com/input-output-hk/cardano-sl/pull/2746) (very unlikely or otherwise we would get this consistently);
3. We are somehow resetting the state accidentally -- for example is sufficient to run the scripts with `--rebuild-db` to cause the node to wipe off completely its wallet DB.

**Comment by parsonsmatt** [comment]
It would be really helpful to have the logs available somewhere for a run of the test -- the current script does not put them anywhere I can see them.

**Comment by akegalj** [comment]
> It would be really helpful to have the logs available somewhere for a run of the test -- the current script does not put them anywhere I can see them.

you mean the very same logs that are on stdout ? Like 
```
[TEST-LOG] Result: Wallet {walId = WalletId "J7rQqaLLHBFPrgJXwpktaMB1B1kQBXAyc2uRSfRPzNVGiv6TdxBzkPNBUWysZZZdhFG9gRy3sQFfX5wfpLbi4XTFGFxTg", walName = "My wallet", walBalance = Coin {getCoin = 37138290651997550}, walHasSpendingPassword = False, walSpendingPasswordLastUpdate = 308128273141538, walCreatedAt = 249785469591062, walAssuranceLevel = NormalAssurance}
[TEST-LOG] Success!
[TEST-LOG] Running: PostWallet
[TEST-LOG] Request: NewWallet {newwalBackupPhrase = <backup phrase>, newwalSpendingPassword = Just <passphrase>, newwalAssuranceLevel = NormalAssurance, newwalName = "Wallet", newwalOperation = CreateWallet}
[TEST-LOG] Mnemonic was already taken!
[TEST-LOG] Success!
[TEST-LOG] Running: GetTransaction
[TEST-LOG] Running: GetWallet
```
?

**Comment by parsonsmatt** [comment]
No, I mean the server's logs. Right now we only know what error the server sends us, or what state -- but we don't know how it got there.

**Comment by akegalj** [comment]
@parsonsmatt 
well, everything should be already in `./logs/<latest_timestamp>/{node3.log,node3.log.pub}`

```
[akegalj@notebook cardano-sl]$ wc -l logs/2018-04-09_195010/{node3.log,node3.log.pub}
  1542 logs/2018-04-09_195010/node3.log
  1422 logs/2018-04-09_195010/node3.log.pub
  2964 total
```

**Comment by parsonsmatt** [comment]
oh heck yeah, thanks! :smile:

**Comment by akegalj** [comment]
@parsonsmatt 
By reviewing current integration tests I have noticed some wallet state management missing. 

@ksaric can you review my `FIXME` notes in this commit https://github.com/input-output-hk/cardano-sl/pull/2728/commits/3db790f3308299822e7db13dfb73261920ebaa66 . I believe we are missing this state management.

**Comment by parsonsmatt** [comment]
I think that asserting equality on state is going to be error prone (or we're going to need to reimplement all of the same wallet logic on the front end, in an ad hoc `StateT` sort of way). Can we rethink our invariants so that they're more flexible?

If the state management gets any trickier, I'm going to want to bring Sqlite in, and that feels like the wrong path to go down :smile:

**Comment by ksaric** [comment]
For now, these changes don't seem to be that difficult - https://github.com/input-output-hk/cardano-sl/commit/3db790f3308299822e7db13dfb73261920ebaa66
I don't see it as difficult, yet.

> I think that asserting equality on state is going to be error prone (or we're going to need to reimplement all of the same wallet logic on the front end, in an ad hoc StateT sort of way). Can we rethink our invariants so that they're more flexible?

It does depend on what do we want to test. If we want the specific behaviour of the wallet and test all its changes, it does seem like this is the way to go - the difference being we don't copy the functionality but we use something similar, similarly the way we use QuickCheck - we check the same thing by using a different method, or we use an inverse.

**Comment by akegalj** [comment]
`LocalAccountDiffers` is still there. I believe there are no more problems except this.

**Comment by spswv** [comment]
mute the thread10.04.2018, 22:08, "Ante Kegalj" <notifications@github.com>:LocalAccountDiffers is still there. I believe there are no more problems except this.

File: wallet-new/integration/Functions.hs [hunk 7]
```diff
@@ -307,8 +307,8 @@ runAction wc action = do
             -- We choose from the existing wallets AND existing accounts.
             account <-  pickRandomElement localAccounts
             let walletId = accWalletId account
-                deleted = filter notSame localAccounts
-                notSame acct =
+                withoutAccount = filter (not . chosenAccount) localAccounts
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: see comment here https://iohk.myjetbrains.com/youtrack/issue/CSL-2445#comment=93-20213 .

File: wallet-new/integration/Main.hs [hunk 8]
```diff
@@ -65,12 +67,16 @@ main = do
 
 initialWalletState :: WalletClient IO -> IO WalletState
 initialWalletState wc = do
+    -- We will have single genesis wallet in intial state that was imported from launching script
```
Comment (akegalj) [inline wallet-new/integration/Main.hs]: see comment here https://iohk.myjetbrains.com/youtrack/issue/CSL-2445#comment=93-20204

File: wallet-new/integration/Functions.hs [hunk 9]
```diff
@@ -211,7 +212,7 @@ runAction wc action = do
 
         UpdateWalletPass -> do
 
-            localWallets <- use nonGenesisWallets
+            localWallets <- use wallets
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: we should be able to update genesis wallet passphrase

File: wallet-new/integration/Functions.hs [hunk 10]
```diff
@@ -270,9 +271,9 @@ runAction wc action = do
 
         GetAccounts   -> do
             -- We choose from the existing wallets AND existing accounts.
-            wallet  <-  pickRandomElement =<< use nonGenesisWallets
+            wallet  <-  pickRandomElement =<< use wallets
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: we should be able to get accounts from genesis wallet

File: wallet-new/integration/Functions.hs [hunk 11]
```diff
@@ -270,9 +271,9 @@ runAction wc action = do
...
             let walletId = walId wallet
-            walletIdIsNotGenesis walletId
+
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: we should be able to get accounts from genesis wallet

File: wallet-new/integration/Functions.hs [hunk 12]
```diff
@@ -290,7 +291,6 @@ runAction wc action = do
             -- We choose from the existing wallets AND existing accounts.
             account <- pickRandomElement =<< use accounts
             let walletId = accWalletId account
-            walletIdIsNotGenesis walletId
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: we should be able to get genesis wallet account

File: wallet-new/integration/Functions.hs [hunk 13]
```diff
@@ -307,12 +307,14 @@ runAction wc action = do
...
                 withoutAccount = filter (not . chosenAccount) localAccounts
                 chosenAccount acct =
-                    accWalletId acct == accWalletId account
-                    && accIndex acct ==  accIndex account
-                accIdx = accIndex account
+                    accWalletId acct == walletId
+                    && accIndex acct ==  accIdx
 
+            -- We don't want to delete accounts in genesis wallets
+            walletIdIsNotGenesis walletId
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: we shouldn't be able to delete account from genesis wallet (we don't want to loose that money)

File: wallet-new/integration/Functions.hs [hunk 14]
```diff
@@ -364,8 +366,9 @@ runAction wc action = do
             -- We choose from the existing wallets AND existing accounts.
             account <-  pickRandomElement localAccounts
             let walletId = accWalletId account
+            walletPass <- use (walletsPass . at walletId)
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: see https://iohk.myjetbrains.com/youtrack/issue/CSL-2462

File: wallet-new/integration/Functions.hs [hunk 15]
```diff
@@ -426,6 +426,7 @@ runAction wc action = do
             accountSource <- pickRandomElement localAccsWithMoney
 
             let accountSourceMoney = accAmount accountSource
+                withoutSourceAddresses = filter (`notElem` accAddresses accountSource) localAddresses
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: see https://iohk.myjetbrains.com/youtrack/issue/CSL-2458

**Comment by akegalj** [comment]
@parsonsmatt 
ok, seems like we are close to be finished. I have successfully run multiple times now without any problems (it finished with success). To be sure, I am going to create a throw-away script that will rerun tests on a clean state for 100+ times. I will log outputs from this runs and will open any new issues based on it.

Expect openning some new issues by end of the day (CET) if there are some remaining problems. 

_ lunch break _

File: wallet-new/integration/Functions.hs [hunk 16]
```diff
@@ -21,467 +25,556 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
...
+import           Pos.Core (getCoin, mkCoin)
 import qualified Pos.Wallet.Web.ClientTypes.Types as V0
 
 import           Error
 import           Types
 
-
--- | The top function that we need to run in order
--- to test the backend.
+newtype RefT s m a
```
Comment (parsonsmatt) [inline wallet-new/integration/Functions.hs]: We were having a problem with state! It would roll back on an `empty` from IO. So I switched to an `IORef` variant that doesn't roll back. This has a custom `MonadState` implementation that writes the IORef.

File: wallet-new/integration/Functions.hs [hunk 17]
```diff
@@ -21,467 +25,556 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
...
+                        sort as `shouldBe` sort bs
+                    LocalTransactionMissing txn txns ->
+                        txns `shouldContain` [txn]
+                    err ->
+                        expectationFailure $ show err
+
+-- | Attempt each action in the list. If an action fails, ignore the
+-- failure and try the next action in the sequence.
+tryAll :: Alternative f => NonEmpty (f a) -> f a
+tryAll (x :| xs) = foldl' (\this next -> (this *> next) <|> next) x xs
```
Comment (parsonsmatt) [inline wallet-new/integration/Functions.hs]: This implementation of `tryAll` does not have the problems of the old one, and the `NonEmpty` constraint makes it even safer!

File: wallet-new/test/Spec.hs [hunk 18]
```diff
@@ -17,3 +26,14 @@ main = hspec $ do
     API.spec
     Swagger.spec
     ReqSpec.spec
+
+    eqProps @WalletAddress
+    eqProps @Address
+    eqProps @Wallet
+    eqProps @Transaction
```
Comment (parsonsmatt) [inline wallet-new/test/Spec.hs]: I'm getting some weird results with `elem` and `shouldBe`. When I don't trust things, I write tests.

File: scripts/test/wallet/integration.sh [hunk 19]
```diff
@@ -12,6 +12,8 @@ scripts=$(dirname "$0")/../..
 sessionName="integration-tests-"`date +%H%M%S`
 tmpSecrets=$scripts/../temp-secrets
 
+SECONDS=$1
```
Comment (parsonsmatt) [inline scripts/test/wallet/integration.sh]: The script now takes a parameter to indicate how long to wait. My system can boot in 20s.

File: wallet-new/integration/Functions.hs [hunk 20]
```diff
@@ -135,8 +161,11 @@ runAction wc action = do
                     <*> pure CreateWallet
 
         GetWallets -> do
-            -- our internal state is less important than
             result  <-  respToRes $ getWallets wc
+--            localWallets <- use wallets
+--            checkInvariant
+--                (length localWallets == length result)
+--                (LocalWalletsDiffers result localWallets)
```
Comment (parsonsmatt) [inline wallet-new/integration/Functions.hs]: I'm commenting this out for the same reason that I've commented out a further problem. Trying to keep all of the details up to date in the local copy of the state is only going to cause massive complexity for questionable benefit. More deterministic unit tests can be used to verify more specific concerns, with a finer control on state.

File: wallet-new/integration/Functions.hs [hunk 21]
```diff
@@ -175,7 +204,7 @@ runAction wc action = do
             -- Modify wallet state accordingly.
             wallets  %= filter (\w -> walId wallet /= walId w)
             accounts .= walletAccounts
-            addresses %= filter (`notElem` concatMap accAddresses walletAccounts)
+            addresses %= filter (`elem` concatMap accAddresses walletAccounts)
```
Comment (parsonsmatt) [inline wallet-new/integration/Functions.hs]: Ugh. Binary logic. We want to keep addresses that belong to the wallets.

File: wallet-new/integration/Functions.hs (hunk 16 above)
Comment (akegalj) [reply wallet-new/integration/Functions.hs]: It is not obvious to me why this was happening (I am a dumb person). Would you give some more pointers (it seems important to understand)

Comment (akegalj) [reply wallet-new/integration/Functions.hs]: Aha, I got it (after some code review) ty

Comment (parsonsmatt) [reply wallet-new/integration/Functions.hs]: Consider how `StateT` does an alternative:

```haskell
StateT f <|> StateT g =
  StateT $ \s -> 
    f s <|> g s
```

If `f` fails with empty, then we pass the same state to `g`. So if we have:

```haskell
foo = do
  bar
  baz
  empty

test = foo <|> quux
```

When we call `foo`, `bar` and `baz` both modify the state. However, `empty` cancels the computation, and `StateT`'s alternative instance passes the state to `quux` that was initially passed to `foo`.

If you layer the transformers differently (eg `MaybeT (StateT ...)`) then the backtracking is different. But `IO` must be on the bottom.

File: wallet-new/src/Cardano/Wallet/Client.hs [hunk 22]
```diff
@@ -120,10 +121,10 @@ data WalletClient m
     } deriving Generic
 
 getAddressIndex :: WalletClient m -> Resp m [WalletAddress]
-getAddressIndex wc = getAddressIndexPaginated wc Nothing Nothing
+getAddressIndex wc = getAddressIndexPaginated wc Nothing $ Just 50
 
 getAccounts :: WalletClient m -> WalletId -> Resp m [Account]
-getAccounts wc wi = getAccountIndexPaged wc wi Nothing Nothing
+getAccounts wc wi = getAccountIndexPaged wc wi Nothing $ Just 50
```
[the remark this answers is not in the document]
Comment (akegalj) [reply wallet-new/src/Cardano/Wallet/Client.hs]: Its WIP https://iohk.myjetbrains.com/youtrack/issue/CSL-2476 " Auto paginate all results in integration tests"

File: wallet-new/integration/Functions.hs [hunk 23]
```diff
@@ -73,10 +73,10 @@ runActionCheck
     -> ActionProbabilities
     -> m WalletState
 runActionCheck walletClient walletState actionProb = do
-    actions <- chooseActions 100 actionProb
+    actions <- chooseActions 10 actionProb
     log $ "Test will run these actions: " <> show (toList actions)
     let client' = hoistClient lift walletClient
-    ws <- execRefT (tryAll (map (runAction client') actions)) walletState
+    ws <- execRefT (tryAll (map (runAction client') actions) <|> pure ()) walletState
```
Comment (akegalj) [inline wallet-new/integration/Functions.hs]: From https://iohk.myjetbrains.com/youtrack/issue/CSL-2480#comment=93-20502 : 

My sentiment is that this doesn't really fix the problem. For example running `[DeleteWallet]` list of actions will now succeed where actually `DeleteWallet` action has been skipped. If someone just made a bug in delete wallet method in API it would make a regression and integration test would be unable to catch it.

With this modification I would at least add a condition for test to succeed is that all actions from list of actions should be successfully run at least once. Do you think this makes sense? Otherwise this test will sometimes succeed with no good reason to succeed.

I can do something like https://github.com/input-output-hk/cardano-sl/blob/267537f31ef597564f1b95e72d52f7e8a40a1dca/wallet-new/integration/Functions.hs#L109 : 
```haskell
   succs <- use successActions
   when (not . null $ ([minBound..maxBound] \\ nub succs) $
      throwM "Test skipped some actions and we don't have a ground to say everything is ok (when not everything has been tested)."
```

Other then that, `tryAll` would (unexpected given its name) unwrup to:
```haskell
((((((action1 *> action2) <|> action2) *> action3) <|> action3) *> action4) <|> action4) ....
```
which shows that only action1 will be tried exactly one time, where all other actions (if certan conditions are met) can be tried two times. This didn't bite us only because:
 * we are running actions in sequence
 * if action is failing, it will be tried again as a next action immediatelly after and most probably it will also fail

That is also the reason why `WalletState.actionsNum` will be greater then list of actions in a list of actions - as some actions are tried and failed multiple times - with no real harm for now (but it might bit us)

So my question is: what should be expected success condition of these tests? Is it:
 * all actions should be run at least once successfully (doesn't have failing invariant)
     - solid condition to me
 * at least one action should be run successfully (doesn't have failing invariant)
     - I find this a weak condition
 * none of actions should have failing invariant
     - this is current condition
     - I find this a weak condition as skipped actions also have no failing conditions (as we didn't test anything)

Summary:
 * what should be successful test condition?
 * do we want to fix `tryAll` which doesn't really work as expected and tries some actions multiple times or we don't care?

File: wallet-new/integration/Main.hs [hunk 24]
```diff
@@ -172,6 +170,34 @@ deterministicTests wc = do
 
             map txId resp `shouldContain` [txId txn]
 
+        it "estimate fees of a well-formed transaction" $ do
+            ws <- (,)
```
Comment (ksaric) [inline wallet-new/integration/Main.hs]: A `do` might be more readable?

File: wallet-new/integration/Main.hs [hunk 25]
```diff
@@ -172,6 +170,34 @@ deterministicTests wc = do
...
+                <$> (randomWallet >>= createWalletCheck)
+                <*> (randomWallet >>= createWalletCheck)
+
+            ((fromAcct, _), (toAcct, toAddr)) <- (,)
```
Comment (ksaric) [inline wallet-new/integration/Main.hs]: A `do` might be more readable here too.

**Comment by parsonsmatt** [comment]
@KtorZ Err! Is this already merged into `develop`?

**Comment by KtorZ** [comment]
@parsonsmatt ... erf... No! I actually renamed the branch to `feature/CBR-184/transaction-integration-tests` as I thought it was a branch you created for that purpose. 
My bad.
